src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,461 @@
+{
+    // Find empty index or duplicate if exists
+    uint16_t subscriptionIndex;
+    uint16_t firstEmptySubscriptionIndex = CHIP_IM_MAX_NUM_SUBSCRIPTIONS; // initialize to out of bounds as "not set"
+    for (subscriptionIndex = 0; subscriptionIndex < CHIP_IM_MAX_NUM_SUBSCRIPTIONS; subscriptionIndex++)
+    {
+        SubscriptionInfo currentSubscriptionInfo;
+        CHIP_ERROR err = Load(subscriptionIndex, currentSubscriptionInfo);
+
+        // if empty and firstEmptySubscriptionIndex isn't set yet, then mark empty spot
+        if ((firstEmptySubscriptionIndex == CHIP_IM_MAX_NUM_SUBSCRIPTIONS) && (err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND))
+        {
+            firstEmptySubscriptionIndex = subscriptionIndex;
+        }
+
+        // delete duplicate
+        if (err == CHIP_NO_ERROR)
+        {
+            if ((subscriptionInfo.mNodeId == currentSubscriptionInfo.mNodeId) &&
+                (subscriptionInfo.mFabricIndex == currentSubscriptionInfo.mFabricIndex) &&
+                (subscriptionInfo.mSubscriptionId == currentSubscriptionInfo.mSubscriptionId))
+            {
+                Delete(subscriptionIndex);
+                // if duplicate is the first empty spot, then also set it
+                if (firstEmptySubscriptionIndex == CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
+                {
+                    firstEmptySubscriptionIndex = subscriptionIndex;
+                }
+            }
+        }
+    }
+
+    // Fail if no empty space
+    if (firstEmptySubscriptionIndex == CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    // Now construct subscription state and save
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    backingBuffer.Calloc(MaxSubscriptionSize());
+    ReturnErrorCodeIf(backingBuffer.Get() == nullptr, CHIP_ERROR_NO_MEMORY);
+
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), MaxSubscriptionSize());
+
+    ReturnErrorOnFailure(Save(writer, subscriptionInfo));
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+
+    ReturnErrorOnFailure(
+        mStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::SubscriptionResumption(firstEmptySubscriptionIndex).KeyName(),
+                                  backingBuffer.Get(), static_cast<uint16_t>(len)));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::Delete(NodeId nodeId, FabricIndex fabricIndex, SubscriptionId subscriptionId)
+{
+    CHIP_ERROR deleteErr = CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+
+    uint16_t count = 0;
+    for (uint16_t subscriptionIndex = 0; subscriptionIndex < CHIP_IM_MAX_NUM_SUBSCRIPTIONS; subscriptionIndex++)
+    {
+        SubscriptionInfo subscriptionInfo;
+        CHIP_ERROR err = Load(subscriptionIndex, subscriptionInfo);
+
+        // delete duplicate
```

Comment:
Is there a duplicate here? or is this just "delete match"?

Maybe comment could be that we do not break on the first delete because there may be duplicates (if they exist is it a bug? Should we log if we delete when count is non-zero?)

Reply:
I mis-read the count usage ... it is the count of still existing subscriptions rather than number of deletions.
I guess then the question is do we care enough to also count deletions and warn if we deleted more than one item?

maybe count should be `remainingSubscriptionsCount` for clarity.

Reply:
Yep you got it. I'll include the renames ("duplicate" and "count")  the next time I push a commit.

Reply:
Not sure if we care about multiple deletions - that shouldn't happen to begin with because we also look for and delete duplicates on Save.

Reply:
Just pushed changes for the renames.